### DISCUSSION TITLE
Linux & Windows containers at the same time?

### DISCUSSION THREAD

**Discussion by brad-jones** [batect/batect#827] [Q&A] [answered]
As per the following statement on the website.

![image](https://user-images.githubusercontent.com/2754772/111407193-99476780-8727-11eb-9fb0-3de7da1a9ade.png)

Can batect run Linux & Windows containers at the same time or is it still tied to the _"Docker Desktop Switch"_?

![image](https://user-images.githubusercontent.com/2754772/111407600-183ca000-8728-11eb-86c9-caf26d8831f1.png)

I have a custom setup with 2 docker demons running, one is Docker Desktop running in Windows Container mode & the other is a standalone docker daemon running in WSL2. I find this useful for legacy Windows stacks that are being incrementally modernized. So some of the stack is Windows & other parts are Linux.

Could batect be configured to talk to different daemons based on the type of container?

**Comment by priorax** [comment]
Is this potentially also a valid use of #78?
Wanting a task runner that works correctly with any legacy projects that run on Windows, but that as you cut things over to Linux containers you can change the implementation in batect from being host to being container based and the same task continues to work?
You do lose out on having version locks of course, but it would still give a seamless developer experience during the upgrade path

**Reply by charleskorn** [reply]
Possibly, but if I understood @brad-jones' question correctly, it sounds like they want to be able to work with both Windows containers and Linux containers as part of the same project, rather than Windows workloads running directly on the host. Have I understood what you were asking @brad-jones?

**Reply by priorax** [reply]
100% agree with your interpretation of the question.
Was more just trying to find a use case for where #78 would be helpful

**Comment by charleskorn** [comment, marked as answer]
> Can batect run Linux & Windows containers at the same time or is it still tied to the "Docker Desktop Switch"?

Batect is limited to whatever the Docker daemon it is talking to supports - so if your local Docker installation is set up for Linux containers, Batect can work with Linux containers, and if your local Docker installation is set up for Windows containers, Batect can work with Windows containers. This is a limitation of Docker unfortunately.

> Could batect be configured to talk to different daemons based on the type of container?

This isn't something that is currently supported, and would add significant complexity to the codebase, so I'd prefer to avoid it unless there's significant demand for it. However, there's nothing stopping you from running different Batect tasks from the same project against different daemons that support different kinds of containers - you'd just need to tell Batect to use the right daemon, either with the `DOCKER_HOST` environment variable or with the [`--docker-host` command line option](https://batect.dev/docs/reference/cli/#--docker-host).

**Reply by brad-jones** [reply]
Yeah right then I need a task runner to run the task runner hahaha
